Force write of local file header when "version needed to extract" changes

edwardneal opened the pull request:
Contributes to #112017.

`aapt2` creates ZIP entries with a "version needed to extract" of 0.0, but sometimes writes these entries with deflate compression (which requires a "version needed to extract" of 2.0.)

When a ZipArchive is opened in Update mode and a new item is added, the central directory is written with the correct value for this field. However, the local file header for existing entries isn't rewritten (because it hasn't changed.) This leads to a mismatch between the local file header values and the central directory header values, which causes Android app builds to fail.

When an existing entry's "version needed to extract" is changed, we force its local file header to be rewritten.

There's a test to cover this. I've also added an entry to the packaged_resources.zip file from the issue and confirmed that the field matches between the local file header and the CD header for all compressed entries in the resultant file.

@carlossanlop @jonathanpeppers

edwardneal committed:
Add failing test

edwardneal committed:
Force local file headers to be rewritten if the version to extract is changed

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -87,9 +86,11 @@ internal ZipArchiveEntry(ZipArchive archive, ZipCentralDirectoryFileHeader cd)
+
+            // Setting CompressionMethod can change the _versionToExtract variable, which can change the value of Changes
+            CompressionMethod = (CompressionMethodValues)cd.CompressionMethod;
+            _compressionLevel = MapCompressionLevel(_generalPurposeBitFlag, CompressionMethod);
```
Is there a motive to move these two together to this location? I'm failing to see one, but maybe I'm wrong. Usually we try to minimize changes.

edwardneal replied:
There is. The setter for `CompressionMethod` calls `VersionToExtractAtLeast`, which now sets `Changes`, so I have to set CompressionMethod after Changes. I've taken another look and reduced the diff slightly though.

carlossanlop replied:
Ok got it, thanks! I would've been fine with your initial change after you shared the motive, but no need to revert it again. This is good too.

carlossanlop commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:
```diff
@@ -907,6 +907,145 @@ public static async Task ZipArchive_InvalidHuffmanData()
+            }
+        }
+
+        private static readonly byte[] s_inconsistentVersionToExtract =
```
Nice, thanks for the comments.

carlossanlop commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:
```diff
@@ -907,6 +907,145 @@ public static async Task ZipArchive_InvalidHuffmanData()
+            0x46, 0xd7, 0x6c, 0x45,
+            // compressed size
+            0x0c, 0x00, 0x00, 0x00,
+            // UNcompressed size
```
```suggestion
            // Uncompressed size
```

edwardneal replied:
Thanks - I've made the casing consistent here, and added an explanation to L923 of the test.

carlossanlop commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:
```diff
@@ -907,6 +907,145 @@ public static async Task ZipArchive_InvalidHuffmanData()
+                Assert.Equal(0x00, originalLocalVersionToExtract);
+                Assert.Equal(0x00, originalCentralDirectoryVersionToExtract);
+
+                updatedStream.Write(s_inconsistentVersionToExtract);
```
A comment along these lines would be good:
```suggestion
                // Writes a full zip archive
                updatedStream.Write(s_inconsistentVersionToExtract);
```

carlossanlop commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:
```diff
@@ -907,6 +907,145 @@ public static async Task ZipArchive_InvalidHuffmanData()
+            0x46, 0xd7, 0x6c, 0x45,
+            // compressed size
+            0x0c, 0x00, 0x00, 0x00,
+            // UNcompressed size
```
```suggestion
            // Uncompressed size
```

carlossanlop reviewed (approved):
Thanks for the fix, @edwardneal! Left some nit comments and a non blocking question, but otherwise the fix LGTM.

> aapt2 creates ZIP entries with a "version needed to extract" of 0.0, but sometimes writes these entries with deflate compression (which requires a "version needed to extract" of 2.0.)

This sentence in the description makes it sound like the bug is in aapt2. My understanding from your explanation in the original issue was that this is a bug in ZipArchive. Or is it a bug in both?

carlossanlop commented:
/azp run runtime-libraries-coreclr outerloop

filipnavara commented:
> My understanding from your explanation in the original issue was that this is a bug in ZipArchive. Or is it a bug in both?

My understanding is that the bug is in both. AAPT2 uses the Deflate algorithm and doesn't correctly set the "version to extract" field. ZipArchive corrects the "version to extract" to a valid value and updates the central directory, but fails to update the locally prefixed header for the corrected entries.

edwardneal committed:
Code review

Adding/correcting comments in test, and slightly reducing the diff in ZipArchiveEntry.

edwardneal commented:
> This sentence in the description makes it sound like the bug is in aapt2. My understanding from your explanation in the original issue was that this is a bug in ZipArchive. Or is it a bug in both?

Thanks @carlossanlop. filipnavara is correct, it's a bug in both. The ZIP file being generated in aapt2 is malformed, but ZipArchive has a bug because it's correcting the central directory header without doing the same for the local file header.

carlossanlop commented:
Lot's of unrelated failures. We rollbacked some helix machine updates due to reported generalized failures. I'll update the branch so the CI runs again.

carlossanlop committed:
Merge branch 'main' into issue-112017

carlossanlop commented:
/azp run runtime-libraries-coreclr outerloop
